rfc/build-system/build-system.md

```diff
@@ -0,0 +1,698 @@
+directory inside this new, larger repo. We will then use
+[Gradle composite builds](https://docs.gradle.org/current/userguide/composite_builds.html)
+to tie these repositories together. There will be no `build.gradle` file in the
+root directory.
+
+You can already test this approach by editing
+[`orca/gradle.properties`](https://github.com/spinnaker/orca/blob/master/gradle.properties)
+and setting `korkComposite` to true. We will simply flip the default behavior,
+so that a composite build is the default and a `korkVersion` can be specified to
+revert to downloading a previously-built artifact.
+
+Only the required dependent modules will be built. For example, `front50` has a
+dependency on `clouddriver`, but building `front50` doesn't require building
+_all_ of `clouddriver`. You just need to build the `clouddriver-google-common`
+module.
+
+Caching works as you'd expect: if you build `orca` followed by `fiat`, you will
+only have to compile each `kork` module once.
+
+This monorepo requires you to change all affected Java code in a single PR; no
+longer will you commit code to `kork` only to find out later it broke
+compilation of a handful of microservices.
+
+**_Alternatives:_** I also wrote a description of a build system that doesn't
+use a monorepo. It's significantly more complicated, but I
+[included it in an appendix](#appendix-an-alternative-multi-repo-workflow).
+
+### CI Builds
+
+For continuous integration, we will use GitHub Actions. A separate job will be
+launched for every sub-project. These jobs will first use
+[a custom action to determine whether or not that project needs to be rebuilt](#requires_build)
+(roughly: did any files in the project or any of its dependencies change?). If
+not, we'll just declare the build a success. If so, we'll do a complete build of
+the project.
+
+Since these jobs run in parallel, building all of them (if required) takes only
+as long as the slowest project (which is `clouddriver`, at about 20 minutes).
+
+**_Alternatives_**: I looked at using a single CI job that would build the
+projects in serial, but it took too long. A change to `kork` required over an
+hour of building.
+
+### Maven Artifact Publishing
+
+Twice a week, via a scheduled GitHub Action, we will run a build that publishes
+Maven artifacts. These aren't used by our builds anymore, but can be used by
+projects that want to depend on a recent, unreleased version of the Java
+microservices. For example, `halyard` will use these artifacts, as will users
+developing plugins for the next version of Spinnaker.
+
+Artifacts will be identified by the branch, commit timestamp (in UTC), and short
+SHA (seven digits). They'll look like this: `2020.4-20200601.193201-198aa21`.
+
+Commit timestamps aren't guaranteed to increase monotonically, so technically
+Maven won't always sort these versions correctly. If, for example, you're
+developing a plugin and add a dependency on the latest version of echo, you may
```

Comment:
For plugins, would we just `require` a Spinnaker version now instead of individual service versions? That would certainly be easier for plugin authors and users.

Reply:
Yep, that's one of the main goals of this (after discussions I had with @link108 and @claymccoy): every component of a Spinnaker release (including deck) has the same version number. Your plugin targets Spinnaker 2021.2 instead of some other obscure version number.

Reply:
I feel like for each service's plugin api we could have a legitimate semver story.

Once we monorepo everything I feel it will be harder for plugin developers to express meaningful version constraint requirements on their plugins other than `service>=the release version where I first wrote this` and plugins breaking on version updates will be a riskier and completely untestable by the core project proposition